### PR TITLE
Interchange linalg generic op to move reduction loops to inner loops

### PR DESCRIPTION
**Pull request by ThomasRaoux**
During dispatchLinalgOnTensor we only tile outer parallel loops so if
the parallel loops are not outer loop we may skip tiling some ops which
will prevent them from being distributed onto several workgroups.

### CODE REVIEW

**Comment by raikonenfnu** [comment]
fwiw LGTM +2 👍

**Review by raikonenfnu — APPROVED** [review]

File: iree/compiler/Dialect/Flow/Transforms/InterchangeGenericOp.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<unsigned> interchange;
+    bool needInterchange = false;
+    unsigned numParallelLoop = genericOp.getNumParallelLoops();
+    if (numParallelLoop == 0) return failure();
+    for (auto iter : llvm::enumerate(genericOp.iterator_types())) {
+      if (isParallelIterator(iter.value())) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/InterchangeGenericOp.cpp:38]: Maybe too aggressive. The Flow only distributed upto 3 loops. Maybe just switch upto 3 dimensions?

File: iree/compiler/Dialect/Flow/Transforms/Passes.td [hunk 2]
```diff
@@ -119,4 +119,10 @@ def StripAndSplatConstantVariables :
   let constructor = "mlir::iree_compiler::IREE::Flow::createStripAndSplatConstantVariablesPass()";
 }
 
+def InterchangeGeneric :
+    Pass<"iree-interchange-generic-op", "FuncOp"> {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/Passes.td:123]: Nit: please make this `iree-flow-interchange-generic-op` (for now). Eventually this pass will move out of Flow.

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Dialect/Flow/Transforms/InterchangeGenericOp.cpp (hunk 1 above)
Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/InterchangeGenericOp.cpp:38]: Do you expect it to be beneficial to restrict it to 3 dimensions? I thought we would always want to move the reduction to the inner loop no matter what, that what the front end has been trying to do. 
I assume we will want to distribute more than 3 dimensions in the future?

Comment (raikonenfnu) [reply iree/compiler/Dialect/Flow/Transforms/InterchangeGenericOp.cpp:38]: I personally would like it if we keep it more general than 3 dimensions :)

Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/InterchangeGenericOp.cpp:38]: I am not convinced that there is a use case to distribute more than 3 dimensions. So my view there is more of a holding pattern. Moving "reduction" to inner most loop might be fine. Just seeing if we have any indicators to aid this decision. I am fine with moving it to the innermost.

Comment (ThomasRaoux) [reply iree/compiler/Dialect/Flow/Transforms/InterchangeGenericOp.cpp:38]: Sounds good, I'll keep it like that for now.

**Pull request merged**